compiler-core/src/ast.rs

```diff
@@ -2330,6 +2333,73 @@ impl TypedExprBitArraySegment {
     }
 }
 
+impl<TypedValue> BitArraySegment<TypedValue, Arc<Type>>
+where
+    TypedValue: HasType + HasLocation + Clone + bit_array::GetLiteralValue,
+{
+    pub(crate) fn check_for_truncated_value(&self) -> Option<BitArraySegmentTruncation> {
+        // Both the size and the value must be two compile-time known constants.
+        let segment_bits = self.bits_size()?;
+        let literal_value = self.value.as_int_literal()?;
+        if literal_value.sign() != Sign::Plus || segment_bits.sign() != Sign::Plus {
```

Comment:
Could also warn when the value is negative and out of range.

If the value is negative then the valid range is `(-(1 << (segment_bits - 1)))..((1 << (segment_bits - 1)) - 1)`

So these would also trigger a warning:

```gleam
<<-129>>
<<-530:10>>
<<-80_000:16>>
```

Note that in some cases the `truncated_into` value will be positive, even though the literal value is negative.

Reply:
Great I'm adding this as well then!